tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,235 @@
+//! Tendermint Websocket event listener client
+
+use crate::{
+    block::Block,
+    net,
+    rpc::response::Wrapper,
+    rpc::Request,
+    rpc::{endpoint::subscribe, Error as RPCError},
+};
+use async_tungstenite::{tokio::connect_async, tokio::TokioAdapter, tungstenite::Message};
+use futures::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::error::Error as stdError;
+
+use tokio::net::TcpStream;
+/// There are only two valid queries to the websocket. A query that subscribes to all transactions
+/// and a query that susbscribes to all blocks.
+pub enum EventSubscription {
+    /// Subscribe to all transactions
+    TransactionSubscription,
+    ///Subscribe to all blocks
+    BlockSubscription,
+}
+
+impl EventSubscription {
+    ///Convert the query enum to a string
+    pub fn as_str(&self) -> &str {
+        match self {
+            EventSubscription::TransactionSubscription => "tm.event='Tx'",
+            EventSubscription::BlockSubscription => "tm.event='NewBlock'",
+        }
+    }
+}
+
+/// Event Listener over websocket.
+/// See: https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+pub struct EventListener {
+    socket: async_tungstenite::WebSocketStream<TokioAdapter<TcpStream>>,
+}
+
+impl EventListener {
+    /// Constructor for event listener
+    pub async fn connect(address: net::Address) -> Result<EventListener, RPCError> {
+        let (host, port) = match address {
+            net::Address::Tcp { host, port, .. } => (host, port),
+            other => {
+                return Err(RPCError::invalid_params(&format!(
+                    "invalid RPC address: {:?}",
+                    other
+                )));
+            }
+        };
+        //TODO This doesn't have any way to handle a connection over TLS
+        let (ws_stream, _unused_tls_stream) =
+            connect_async(&format!("ws://{}:{}/websocket", host, port)).await?;
+        Ok(EventListener { socket: ws_stream })
+    }
+    //TODO Have a query type instead of a string
+    /// Subscribe to event query stream over the websocket
+    pub async fn subscribe(&mut self, query: EventSubscription) -> Result<(), RPCError> {
+        self.socket
+            .send(Message::text(
+                subscribe::Request::new(query.as_str().to_owned()).into_json(),
+            ))
+            .await?;
+        Ok(())
+    }
+
+    /// Get the next event from the websocket
+    pub async fn get_event(&mut self) -> Result<Event, RPCError> {
+        let msg = self
+            .socket
+            .next()
+            .await
+            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+        match serde_json::from_str::<JsonRPCBlockResult>(&msg.to_string()) {
+            Ok(data) => {
+                if let Some(block_result) = data.0.result {
```

Comment:
Why not use the methods to access these fields? (`into_result()`)

Reply:
This would also account for malformatted responses (e.g. empty ones ...) and remove the panic below too.
I'll push sth. shortly.

Reply:
b11ea9a